pwndbg/aglib/regs.py

```diff
@@ -0,0 +1,296 @@
+"""
+Reading register value from the inferior, and provides a
+standardized interface to registers like "sp" and "pc".
+"""
+
+from __future__ import annotations
+
+import ctypes
+import re
+import sys
+from types import ModuleType
+from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import Generator
+from typing import List
+from typing import Tuple
+from typing import cast
+
+import pwndbg
+import pwndbg.aglib.arch
+import pwndbg.aglib.proc
+import pwndbg.aglib.qemu
+import pwndbg.aglib.remote
+import pwndbg.aglib.typeinfo
+import pwndbg.lib.cache
+from pwndbg.dbg import EventType
+from pwndbg.lib.regs import BitFlags
+from pwndbg.lib.regs import RegisterSet
+from pwndbg.lib.regs import reg_sets
+
+
+@pwndbg.lib.cache.cache_until("stop", "prompt")
+def regs_in_frame(frame: pwndbg.dbg_mod.Frame) -> pwndbg.dbg_mod.Registers:
+    return frame.regs()
+
+
+@pwndbg.aglib.proc.OnlyWhenRunning
+def get_register(
+    name: str, frame: pwndbg.dbg_mod.Frame | None = None
+) -> pwndbg.dbg_mod.Value | None:
+    if frame is None:
+        frame = pwndbg.dbg.selected_frame()
+    assert (
+        frame is not None
+    ), "pwndbg.dbg.selected_frame() should never return None when marked with @OnlyWhenRunning"
+
+    regs = regs_in_frame(frame)
+
+    return regs.by_name(name) or regs.by_name(name.upper())
+
+
+@pwndbg.aglib.proc.OnlyWhenQemuKernel
+@pwndbg.aglib.proc.OnlyWhenRunning
+def get_qemu_register(name: str) -> int | None:
+    out = pwndbg.dbg.selected_inferior().send_monitor("info registers")
+    match = re.search(rf'{name.split("_")[0]}=\s+([\da-fA-F]+)\s+([\da-fA-F]+)', out)
+
+    if match:
+        base = int(match.group(1), 16)
+        limit = int(match.group(2), 16)
+
+        if name.endswith("LIMIT"):
+            return limit
+        else:
+            return base
+
+    return None
+
+
+# We need to manually make some ptrace calls to get fs/gs bases on Intel
+PTRACE_ARCH_PRCTL = 30
+ARCH_GET_FS = 0x1003
+ARCH_GET_GS = 0x1004
+
+gpr: Tuple[str, ...]
+common: List[str]
+frame: str | None
+retaddr: Tuple[str, ...]
+flags: Dict[str, BitFlags]
+extra_flags: Dict[str, BitFlags]
+stack: str
+retval: str | None
+all: List[str]
+changed: List[str]
+fsbase: int
+gsbase: int
+current: RegisterSet
+fix: Callable[[str], str]
+items: Callable[[], Generator[Tuple[str, Any], None, None]]
+previous: Dict[str, int]
+last: Dict[str, int]
+pc: int | None
+
+
+class module(ModuleType):
```

Comment:
(Random comment: This whole module class design with global attrs is weird. Should probably move off from that in future)